packages/react-reconciler/src/ReactChildFiber.js

```diff
@@ -1141,34 +1158,63 @@ function ChildReconciler(shouldTrackSideEffects) {
       // TODO: If key === null and child.key === null, then this only applies to
       // the first item in the list.
       if (child.key === key) {
-        if (
-          child.tag === Fragment
-            ? element.type === REACT_FRAGMENT_TYPE
```

Comment:
I don't think removing this check is correct.

As a result, `Fragment -> SomethingElse` change won't reset the child state.

Here's a small case to show it:

```js
  fit('should reset state when reconciling from fragment to another type', async () => {
    function Foo({children}) {
      return children;
    }

    function App({step}) {
      let Wrapper = step === 0 ? React.Fragment : Foo;
      return (
        <>
          <Wrapper>
            <Child />
          </Wrapper>
        </>
      )
    }

    let _setChildState;
    function Child() {
      const [x, setX] = React.useState('hello')
      _setChildState = setX;
      return x;
    }

    await ReactNoop.act(async () => {
      ReactNoop.render(<App step={0} />);
    });
    expect(ReactNoop.getChildren()[0].text).toBe('hello');

    // Change the child state
    await ReactNoop.act(async () => {
      _setChildState('goodbye');
    });
    expect(ReactNoop.getChildren()[0].text).toBe('goodbye');
    // Ensure it stays
    await ReactNoop.act(async () => {
      _setChildState('goodbye');
    });
    expect(ReactNoop.getChildren()[0].text).toBe('goodbye');

    // This should reset the child state.
    await ReactNoop.act(async () => {
      ReactNoop.render(<App step={1} />);
    });
    expect(ReactNoop.getChildren()[0].text).toBe('hello');
  });
```

Reply:
If you add a test for this, probably worth making it switch back to Frag too just for completeness.

Reply:
As a fix I think you want to wrap the Fragment case in `element.type === REACT_FRAGMENT_TYPE`. (And add a `break` to it?)

Reply:
Actually I think a break would be insufficient? Since you want to go to the fallthrough case. But you don't want to fall through to the Chunk (or any other case we add in the middle). So I think this needs more restructuring.

Reply:
Yea this is just an oversight in the copy paste. I was wondering if we can just get rid of it though and have the elementType comparison.

Reply:
Might be something to consider before landing the JSX changes @lunaruan is working on.